drivers/can/can_sam0.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2022 Vestas Wind Systems A/S
+ * Copyright (c) 2021 Alexander Wachter
+ * Copyright (c) 2022 Kamil Serwus
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+
+#include "can_mcan.h"
+
+LOG_MODULE_REGISTER(can_sam0, CONFIG_CAN_LOG_LEVEL);
+
+#define DT_DRV_COMPAT atmel_sam0_can
+
+struct can_sam0_config {
+	void (*config_irq)(void);
+	const struct pinctrl_dev_config *pcfg;
+	volatile uint32_t *mclk;
+	uint32_t mclk_mask;
+	uint16_t gclk_core_id;
+	int divider;
+};
+
+struct can_sam0_data {
+	struct can_mcan_msg_sram msg_ram;
+};
+
+void can_sam0_line_x_isr(const struct device *dev)
+{
+	can_mcan_line_0_isr(dev);
+	can_mcan_line_1_isr(dev);
+}
+
+static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_cfg = dev->config;
+	const struct can_sam0_config *sam_cfg = mcan_cfg->custom;
+
+	*rate = SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / (sam_cfg->divider);
```

Comment:
Thanks!
I am not really familiar with Atmel SoCs, but in `can_sam0_clock_enable()` you are setting `GCLK_GENCTRL_SRC_OSC48M` as clock source for the _clock generator 1_, which I assume is a fixed 48MHz oscillator. But here you are using `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` which depends on the selected sysclock frequency and might be changed. Shouldn't you set it to 48MHz or a macro representing the 48MHz here?

Also, the clock generator 1 is part of the Generic Clock Controller peripheral, so having it's divider as a property of the CAN is not ideal. Maybe @nandojve can weight in? Can we introduce that as a temporary solution until a generic clock controller driver is available for the series?

Reply:
I'm not expert on CAN but as far I know clock should be very accurate. I think there is no problem to use `OSC48M` as base clock with proper factory (MCHP) calibration. What could be an issue is that DPLL is not enabled at this moment and that, on my view, should be the GCLKGEN reference for CAN. Usually a `XOSC[32K]` are a better reference for DPLL but you can use `OSC48M` -> `GLKC` -> `DPLL input ref` that will be far better than only `GLKC` (in theory).

Usually without a good clock base CAN may suffer when temperature changes, for instance.

Reply:
What is recommended clock source/frequency for the MCAN block according to the datasheet?

Reply:
@henrikbrixandersen documentation doesn't specify recomended source of clock.

> Clocks
> An AHB clock (CLK_CAN_AHB) is required to clock the CAN. This clock can be configured in the Main Clock
peripheral (MCLK) before using the CAN, and the default state of CLK_CAN_AHB can be found in the
MCLK.AHBMASK register.
> A generic clock (GCLK_CAN) is required to clock the CAN. This clock must be configured and enabled in the generic
clock controller before using the CAN.
> This generic clock is asynchronous to the bus clock (CLK_CAN_AHB). Due to this asynchronicity, writes to certain

Datasheet only specify that MCAN has to be connected to GLCK.

Reply:
I my first comment I wanted to highlight that:
- You are using GLCK1 as clock source for the CAN peripheral.
- For rate calculation in line 46 you use GLCK0 frequency divided by the GLCK1 divider
- But GLCK1 is feed by OSC48M directly and not in series after GCLK0.
- This will work now as the system clock(GLCK0) is also fed by OSC48M and not divided. But as soon as that changes, chances are high that the CAN driver is overlooked and rate calculation is wrong.
- But if Atmel maintainers are fine with it I won't block.

Reply:
@nandojve Just to make sure - are you okay with the above limitations?

Reply:
I'm also not convinced that `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` is appropriate here...

Reply:
I changed reference from `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` to `SOC_ATMEL_SAM0_OSC48M_FREQ_HZ`
I hope that it is ok for you :)